GS-729b: A cleaner fix changes code in the lower layer.

YuchengZhuQH opened the pull request:
## Description:
<!-- Describe your changes in detail / numbered list of distinct changes for reference -->
- Fixed the bug where a user created with an empty enrolkey gets enrolled in all self-enrollable courses with `password = ''`.
- Also works for a course allowing group enrolment, whether or not the group enrolment key is empty.
- Similar to the dirty fix online, but moved to a lower layer of function.
- JorritDeKlerk's fix also prevents an illegal user from logging in regardless of whether they enrolled in a course. (We can't trigger these conditions with our log-in GUI.)


## Checklist before requesting a review:
<!-- Remove non-applicable items e.g. epic sub-issue points -->
<!-- Add an 'x' inside the brackets to check the item -->

- [x] I have performed a self review of my code.
- [x] My code follows the [Moodle Coding Style](https://moodledev.io/general/development/policies/codingstyle).
- [x] Code has been commented, particularly in hard-to-understand areas.

YuchengZhuQH committed:
GS-729b: A cleaner fix changes code in the lower layer.

NickLambellGCH commented on auth.php:
```diff
@@ -351,7 +351,12 @@ private function email_confirmation(string $email = '') {
     private function get_enrol_plugins(moodle_database $db, string $enrolkey) : array {
         // Password is the Enrolment key that is specified in the Self enrolment instance.
-        $enrolplugins = $db->get_records('enrol', ['enrol' => 'self', 'password' => $enrolkey]);
+        if ($enrolkey != '') {
+            $enrolplugins = $db->get_records('enrol', ['enrol' => 'self', 'password' => $enrolkey]);
+        } else {
+            // Avoid the bug that a user created with an empty enrolkey gets enrolled in all courses with `password=''`
+            $enrolplugins = [];
+        }
```
Instead of the `else` content setting `$enrolplugins` to an empty array you should just have that be the default for a variable declaration before if main `if` line.

NickLambellGCH reviewed (changes requested):
Rather than placing the enrol key empty check in a subfunction where it wouldn't immediately be clear that filtering is happening. It might make more sense to have a simple check at the start of `auth.php::enrol_user()` to see if the enrol key is empty and just return straight away to completely skip the enrolment logic.
This has the added side effect / bonus that it addresses a secondary issue from a comment on the upstream issue related to group enrolment keys since that logic is skipped too.

YuchengZhuQH committed:
GS-729b: Made the else block the default value.

YuchengZhuQH commented:
> Rather than placing the enrol key empty check in a subfunction where it wouldn't immediately be clear that filtering is happening. It might make more sense to have a simple check at the start of `auth.php::enrol_user()` to see if the enrol key is empty and just return straight away to completely skip the enrolment logic. This has the added side effect / bonus that it addresses a secondary issue from a comment on the upstream issue related to group enrolment keys since that logic is skipped too.

Good point. I'll test the group enrolment keys issue. I think the main issue is whether a user with an empty enrol key skips all the logics, or they just need to avoid matching the `$db->get_records('enrol', ['enrol' => 'self', 'password' => $enrolkey])` records.

YuchengZhuQH committed:
Revert "GS-729b: A cleaner fix changes code in the lower layer."

This reverts commit f7e2c67a0550f7135e2358eb0fa4d7ad71bc593a. Because the fix doesn't catch the group enrolment bug.

YuchengZhuQH committed:
Revert "GS-729b: Made the else block the default value."

This reverts commit 93c3e2e684979e59a5b3ce67c1a0e64c0be10a9f. Because the fix doesn't catch the group enrolment bug.

YuchengZhuQH committed:
GS-729b: Checked `$enrolkey` is empty in `enrol_user()`

YuchengZhuQH committed:
GS-729b: JorritDeKlerk's fix also prevents an illegal user from logging in.

Regardless of whether they enrolled in a course. We can't trigger these conditions with our log in GUI.

YuchengZhuQH committed:
GS-729b: Fixed grammar in comments.

YuchengZhuQH committed:
GS-729b: Clarified comment.
